File: Spiral Matrix - GFG/spiral-matrix.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution{

	public:
	int findK(vector<vector<int>> &matrix, int n, int m, int k)
    {
        vector<int> res;

        int rs = 0, re = matrix.size() - 1, cs = 0, ce = matrix[0].size() - 1;
        while(rs <= re && cs <= ce)
        {
            for(int i = cs; i <= ce; i++) 
             res.push_back(matrix[rs][i]);
            rs++;
            for(int i = rs; i <= re; i++) 
             res.push_back(matrix[i][ce]);
            ce--;
            for(int i = ce; i >= cs && rs <= re; i--) 
             res.push_back(matrix[re][i]);
            re--;
            for(int i = re; i >= rs && cs <= ce; i--) 
             res.push_back(matrix[i][cs]);
            cs++;
        }
        
        return res[k-1];
    }

};


//{ Driver Code Starts.

int main()
{
    int T;
    cin>>T;
  
    while(T--)
    {
        int n,m;
        int k=0;
        //cin>>k;
        cin>>n>>m>>k;
        vector<vector<int>> a(n, vector<int>(m, 0));
        
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                cin>>a[i][j];
            }
        }

        Solution obj;

        cout<< obj.findK(a, n, m, k) << "\n";
        
       
    }
}
// } Driver Code Ends

File: 54-spiral-matrix/spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        ans = []
        
        sr, er = 0, len(matrix)-1
        sc, ec = 0, len(matrix[0])-1

        while(sr <= er and sc <= ec):
            for j in range(sc, ec+1):
                ans.append(matrix[sr][j])
            sr += 1
            
            for i in range(sr, er+1):
                ans.append(matrix[i][ec])
            ec -= 1

            if sr <= er:
                for j in range(ec, sc-1, -1):
                    ans.append(matrix[er][j])
                er -= 1

            if sc <= ec:
                for i in range(er, sr-1, -1):
                    ans.append(matrix[i][sc])
                sc += 1

        return ans
            
            
            